reader.go

```diff
@@ -0,0 +1,121 @@
+package recordio
+
+import (
+	"io"
+	"log"
+)
+
+// Index consists offsets and sizes of the consequetive chunks in a RecordIO file.
+type Index struct {
+	chunkOffsets []int64
+	chunkLens    []uint32
+	numRecords   int // the number of all records in a file.
+}
+
+// LoadIndex scans the file and parse chunkOffsets, chunkLens, and len.
+func LoadIndex(r io.ReadSeeker) (*Index, error) {
+	f := &Index{}
+	offset := int64(0)
+	var e error
+	var hdr *Header
+
+	for {
+		if hdr, e = parseHeader(r); e != nil {
+			break
+		} else {
+			f.chunkOffsets = append(f.chunkOffsets, offset)
+			f.chunkLens = append(f.chunkLens, hdr.numRecords)
+			f.numRecords += int(hdr.numRecords)
+			offset, e = r.Seek(int64(hdr.compressedSize), io.SeekCurrent)
+			if e != nil {
+				break
+			}
+		}
+	}
+
+	if e == io.EOF {
+		return f, nil
+	}
+	return nil, e
+}
+
+// NumRecords returns the total number of records in a RecordIO file.
+func (r *Index) NumRecords() int {
+	return r.numRecords
+}
+
+// Locate returns the index of chunk that contains the given record,
+// and the record index within the chunk.  It returns (-1, -1) if the
+// record is out of range.
+func (r *Index) Locate(recordIndex int) (int, int) {
+	sum := 0
+	for i, l := range r.chunkLens {
+		sum += int(l)
+		if recordIndex < sum {
+			return i, recordIndex - sum + int(l)
+		}
+	}
+	return -1, -1
+}
+
+// Scanner scans records in a specified range within [0, numRecords).
+type Scanner struct {
+	reader          io.ReadSeeker
+	index           *Index
+	start, end, cur int
+	chunkIndex      int
+	chunk           *Chunk
+	err             error
+}
+
+// NewScanner creates a scanner that sequencially reads records in the
+// range [start, start+len).
+func NewScanner(r io.ReadSeeker, index *Index, start, len int) *Scanner {
+	if start < 0 {
+		start = 0
+	}
+	if len < 0 || start+len >= index.NumRecords() {
+		len = index.NumRecords() - start
+	}
+
+	return &Scanner{
+		reader:     r,
+		index:      index,
+		start:      start,
+		end:        start + len,
+		cur:        start - 1, // The intial status required by Scan.
+		chunkIndex: -1,
+		chunk:      &Chunk{},
+	}
+}
+
+// Scan moves the cursor forward for one record and loads the chunk
+// containing the record if not yet.
+func (s *Scanner) Scan() bool {
+	s.cur++
+
+	if s.cur >= s.end {
+		s.err = io.EOF
+	} else {
+		if ci, _ := s.index.Locate(s.cur); s.chunkIndex != ci {
+			s.chunkIndex = ci
+			s.chunk, s.err = parseChunk(s.reader, s.index.chunkOffsets[ci])
+		}
+	}
+
+	return s.err == nil
+}
+
+// Record returns the record under the current cursor.
+func (s *Scanner) Record() []byte {
+	ci, ri := s.index.Locate(s.cur)
+	if s.chunkIndex != ci {
+		log.Fatalf("Must call Scan before Record")
```

Comment:
Please do not panic when the application is not in an unrecoverable state. Please see bufio.Text() (also a scanner) for reference: https://golang.org/src/bufio/scan.go?s=4262:4293#L96, or here: https://golang.org/src/text/scanner/scanner.go?s=16921:16957#L652
I think returning nil here is reasonable.
Btw, from my understanding, logging inside a library does not make much sense (should report error if need logging) because it "pollutes" stdout. I have never see it in golang official library, but not 100% sure, so I create this question: http://stackoverflow.com/questions/43708125/should-golang-library-do-logging-instead-of-return-error